core/schema/modulesource.go

```diff
@@ -149,20 +148,47 @@ type parsedRefString struct {
 	modPath    string
 	modVersion string
 	hasVersion bool
-	isGitHub   bool
 	kind       core.ModuleSourceKind
+	repoRoot   *vcs.RepoRoot
+}
+
+// interface used for host interaction mocking
+type BuildkitClient interface {
+	StatCallerHostPath(ctx context.Context, path string, followLinks bool) (*types.Stat, error)
 }
 
-func parseRefString(refString string) parsedRefString {
+// parseRefString parses a ref string into its components
+// New heuristic:
+// - stat folder to see if dir is present
+// - if not, try to isolate root of git repo from the ref
+// - if nothing worked, fallback as local ref, as before
+func parseRefString(ctx context.Context, bk BuildkitClient, refString string) parsedRefString {
 	var parsed parsedRefString
 	parsed.modPath, parsed.modVersion, parsed.hasVersion = strings.Cut(refString, "@")
-	parsed.isGitHub = strings.HasPrefix(parsed.modPath, "github.com/")
 
-	if !parsed.hasVersion && !parsed.isGitHub {
-		parsed.kind = core.ModuleSourceKindLocal
+	// We do a stat in case the mod path github.com/username is a local directory
+	stat, err := bk.StatCallerHostPath(ctx, parsed.modPath, false)
+	if err == nil {
+		if !parsed.hasVersion && stat.IsDir() {
+			parsed.kind = core.ModuleSourceKindLocal
+			return parsed
+		}
+	}
+
+	// we try to isolate the root of the git repo
+	repoRoot, err := vcs.RepoRootForImportPath(parsed.modPath, false)
+	if err == nil && repoRoot != nil && repoRoot.VCS != nil && repoRoot.VCS.Name == "Git" {
+		parsed.kind = core.ModuleSourceKindGit
+		parsed.repoRoot = repoRoot
 		return parsed
 	}
-	parsed.kind = core.ModuleSourceKindGit
+
+	// log warning to hint that the remote ref fallbacked as a local source kind
+	if err != nil {
+		slog.Warn("ref %s has not been parsed as a git remote: %v", refString, err)
+	}
```

Comment:
Hm, so I think we'll only hit this when later this will fail right? We're requesting a module source that can't be resolved to a git repo, and wasn't a local directory.

So this will fail later, I think it's fine to silently fail here IMO.

Reply:
Removed the slog accordingly, as it will fail later you are right